File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/ActivateSmartCardCommand.java
package com.worldventures.dreamtrips.wallet.service.command;

import com.worldventures.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCard;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCard;
import com.worldventures.dreamtrips.wallet.service.SmartCardInteractor;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import rx.Observable;

@CommandAction
public class ActivateSmartCardCommand extends Command<SmartCard> implements InjectableAction {

   private final static long DEFAULT_AUTO_CLEAR_DELAY_CLEAR_RECORDS_MINS = 2 * 60 * 24;

   @Inject SmartCardInteractor smartCardInteractor;

   @Override
   protected void run(CommandCallback<SmartCard> callback) throws Throwable {
      smartCardInteractor.activeSmartCardPipe()
            .createObservableResult(new ActiveSmartCardCommand(sc ->
                  ImmutableSmartCard.builder().from(sc).cardStatus(SmartCard.CardStatus.ACTIVE).build()))
            .map(Command::getResult)
            .doOnNext(smartCard -> setDefaultValues())
            .subscribe(callback::onSuccess, callback::onFail);
   }

   private Observable<SetAutoClearSmartCardDelayCommand> setDefaultValues() {
      return smartCardInteractor.autoClearDelayPipe()
            .createObservableResult(new SetAutoClearSmartCardDelayCommand(DEFAULT_AUTO_CLEAR_DELAY_CLEAR_RECORDS_MINS));
   }
}

File: wallet/src/main/java/com/worldventures/wallet/service/command/ActivateSmartCardCommand.kt
package com.worldventures.wallet.service.command

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.domain.entity.CardStatus
import com.worldventures.wallet.domain.entity.SmartCard
import com.worldventures.wallet.service.SmartCardInteractor
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import rx.Observable
import javax.inject.Inject

@CommandAction
class ActivateSmartCardCommand : Command<SmartCard>(), InjectableAction {

   @Inject lateinit var smartCardInteractor: SmartCardInteractor

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<SmartCard>) {
      smartCardInteractor.activeSmartCardPipe()
            .createObservableResult(ActiveSmartCardCommand { it.copy(cardStatus = CardStatus.ACTIVE) })
            .map { it.result }
            .doOnNext { setDefaultValues() }
            .subscribe({ callback.onSuccess(it) }, { callback.onFail(it) })
   }

   private fun setDefaultValues(): Observable<SetAutoClearSmartCardDelayCommand> {
      return smartCardInteractor.autoClearDelayPipe()
            .createObservableResult(SetAutoClearSmartCardDelayCommand(DEFAULT_AUTO_CLEAR_DELAY_CLEAR_RECORDS_MINS))
   }

   companion object {

      private val DEFAULT_AUTO_CLEAR_DELAY_CLEAR_RECORDS_MINS = (2 * 60 * 24).toLong()
   }
}
